widget/IntersectionBetweenModelErrors.tsx

```diff
@@ -238,52 +179,164 @@ class IntersectionBetweenModelErrors extends Component<IntersectionBetweenModelE
+        let Ra;
+        let Rb;
+        let Aab;
+        let totalArea = 1;
+        if (a >= b) {
+          totalArea = (50 * 50 * 3.14) / aProportion;
+          Ra = 50;
+          Rb = Math.sqrt(bProportion * totalArea / 3.14);
+          Aab = abProportion * totalArea;
+        } else {
+          totalArea = (50 * 50 * 3.14) / bProportion;
+          Rb = 50;
+          Ra = Math.sqrt(aProportion * totalArea / 3.14);
+          Aab = abProportion * totalArea;
+        }
 
-            bringToTop("Ra");
+        // This function calcuates the area of overlap
+        // of two circles of radii r1 and r2 whose
+        // centers are separated by a distance d.
+        function areaOverlap(r1, r2, dist) {
+          var r = Math.min(r1, r2);
+          var R = Math.max(r1, r2);
+          if (dist == 0) {
+            return (3.14 * Math.pow(r, 2));
+          } else if (dist >= (R + r)) {
+            return 0;
+          }
+
+          var sectorAreas = Math.pow(r, 2) * Math.acos((Math.pow(dist, 2) +
+            Math.pow(r, 2) - Math.pow(R, 2)) / (2 * dist * r)) +
+            Math.pow(R, 2) * Math.acos((Math.pow(dist, 2) + Math.pow(R, 2) - Math.pow(r, 2)) / (2 * dist * R));
+
+          var triangleAreas = 1/2 * Math.sqrt((-dist + r + R) * (dist + r - R) * (dist - r + R) * (dist + r + R));
+          var overlapArea = sectorAreas - triangleAreas;
+
+          return overlapArea;
+        }
 
-            d3.select(this).attr("stroke-width", "2px");
-          })
-          .on("mousemove", function() {
-            var vennDiagramPlot = document.getElementById("venndiagramplot");
-            var coords = d3.mouse(vennDiagramPlot);
-            tooltip.style("left", `${coords[0] - (margin.left + margin.right)/2}px`)
-              .style("top", `${coords[1] - (margin.top + margin.bottom)/2}px`);
-          })
-          .on("mouseout", function() {
-            tooltip.style("opacity", 0);
-            d3.select(this).attr("stroke-width", "1px");
-          })
-          .on("click", function() {
-            _this.props.filterByInstanceIds(errorPartition[0]);
-          });
+        // In order to decide the distance d that the circles
+        // Need to be from each other, we use the bisection method
+        // to search for the point at which the difference between
+        // The area of overlap between the circles and the actual
+        // are of the intersection is minimal.
+        function aIntersection(dist) {
+          return (areaOverlap(Ra, Rb, dist) - Aab);
+        }
 
-      areas.append("circle")
-          .attr("r", Rb)
-          .attr('transform',
-              "translate(" +
-              xCenter2 + "," +
-              yCenter + ")")
-          .attr("fill", "rgba(206, 160, 205, 0.8)")
+        var r = Math.min(Ra, Rb);
+        var R = Math.max(Ra, Rb);
+
+        // We perform the bisection search between the two extreme values
+        let d = bisect(aIntersection, (r + R - 0.00001), (R - r + 0.00001));
```

Comment:
Sometimes bisect fails for me with the error "f(a) and f(b) must evaluate to opposite signs". Is that never supposed to happen? Is there a way to make the code more robust so the widget doesn't crash if bisect throws an error?

Reply:
Good question. I noticed this too, and am working on a fix.

Basically the case when this happens is likely one of the two extremal cases, i.e. empty intersection or an intersection that has one set wholly included in the other.

It should be possible to fix this by checking ahead of time and not running the bisect procedure when we are in either of those regimes.